File: typescript/0735-asteroid-collision.ts
function asteroidCollision(asteroids: number[]): number[] {
    const stack: number[] = [];

    for (let a of asteroids) {
        while (stack && a < 0 && stack[stack.length - 1] > 0) {
            let diff = a + stack[stack.length - 1];
            if (diff < 0) {
                stack.pop();
            } else if (diff > 0) {
                a = 0;
            } else {
                stack.pop();
                a = 0;
            }
        }
        if (a === 0) continue;
        stack.push(a);
    }

    return stack;
}

File: swift/0735-asteroid-collision.swift
class Solution {
    func asteroidCollision(_ asteroids: [Int]) -> [Int] {
        var stack = [Int]()

        for var a in asteroids {
            while !stack.isEmpty && a < 0 && stack.last! > 0 {
                let diff = a + stack.last!
                if diff < 0 {
                    stack.popLast()
                } else if diff > 0 {
                    a = 0
                } else {
                    a = 0
                    stack.popLast()
                }
            }

            if a != 0 {
                stack.append(a)
            }
        }

        return stack
    }
}